ssl/quic/quic_route.c

```diff
@@ -0,0 +1,326 @@
+ ****************************************************************************/
+
+/*
+ * Generate a new local connection ID
+ */
+int ossl_quic_conn_id_generate(QUIC_CONN_ID *cid)
+{
+    cid->id_len = QUIC_CONN_ID_LEN;
+    return RAND_bytes(cid->id, cid->id_len) > 0;
+}
+
+int ossl_quic_conn_id_set(QUIC_CONN_ID *cid, const unsigned char *id,
+                          unsigned int id_len)
+{
+    if (id_len > sizeof(cid->id))
+        return 0;
+#ifndef NDEBUG
+    if (id_len < sizeof(cid->id))
+        memset(cid->id + id_len, 0, sizeof(cid->id) - id_len);
+#endif
+    memcpy(cid->id, id, cid->id_len = id_len);
+    return 1;
+}
+
+/****************************************************************************
+ * Hash table lookup functions                                              *
+ ****************************************************************************/
+
+static unsigned long id_hash(const QUIC_CONN_ID *id)
+{
+    /*
+     * Naive implemenation of Bernstein's DJBX33A hash, starting with the length
+     * byte, followed by the ID itself.  For local ID's we could do better because
+     * we know the fixed length but it's unlikely to be worthwhile.
+     */
+    const unsigned int len = id->id_len;
+    unsigned long h = len;
+    unsigned int i;
+
+    for (i = 0; i < len; i++)
+        h = 33 * h + id->id[i];
+    return h;
+}
+
+static int id_cmp(const QUIC_CONN_ID *a, const QUIC_CONN_ID *b)
+{
+    if (a->id_len != b->id_len)
+        return a->id_len > b->id_len ? 1 : -1;
+    return memcmp(a->id, b->id, a->id_len);
+}
+
+static unsigned long local_id_hash(const QUIC_ROUTE *a)
+{
+    return id_hash(&a->local);
+}
+
+static int local_id_cmp(const QUIC_ROUTE *a, const QUIC_ROUTE *b)
+{
+    return id_cmp(&a->local, &b->local);
+}
+
+static unsigned long remote_id_hash(const QUIC_ROUTE *a)
+{
+    return id_hash(&a->remote);
+}
+
+static int remote_id_cmp(const QUIC_ROUTE *a, const QUIC_ROUTE *b)
+{
+    return id_cmp(&a->remote, &b->remote);
+}
+
+/****************************************************************************
+ * Local ID's (global)                                                      *
+ ****************************************************************************/
+
+static LHASH_OF(QUIC_ROUTE) *local_id_tbl;
+static CRYPTO_ONCE local_connection_id_init = CRYPTO_ONCE_STATIC_INIT;
+static CRYPTO_RWLOCK *local_connection_lock;
+
+static void local_connection_id_free(void)
+{
+    lh_QUIC_ROUTE_free(local_id_tbl);
+    local_id_tbl = NULL;
+    CRYPTO_THREAD_lock_free(local_connection_lock);
+    local_connection_lock = NULL;
+}
+
+DEFINE_RUN_ONCE_STATIC(local_connection_id_do_init)
+{
+    local_id_tbl = lh_QUIC_ROUTE_new(&local_id_hash, &local_id_cmp);
+    if (local_id_tbl == NULL)
+        return 0;
+    local_connection_lock = CRYPTO_THREAD_lock_new();
+    if (local_connection_lock == NULL) {
+        lh_QUIC_ROUTE_free(local_id_tbl);
+        return 0;
+    }
+    OPENSSL_atexit(&local_connection_id_free);
```

Comment:
This should instead be called from `OPENSSL_cleanup()`. Some applications disable atexit handlers and call `OPENSSL_cleanup` directly.

Reply:
`ssl_library_stop()` not `OPENSSL_cleanup()`, but done.